crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+        }
+    }
+
+    /// An arbitrary component error. Possibly related to reflection.
+    #[must_use]
+    pub fn component_error(error: anyhow::Error) -> Self {
+        Self {
+            code: error_codes::COMPONENT_ERROR,
+            message: error.to_string(),
+            data: None,
+        }
+    }
+
+    /// An arbitrary internal error.
+    #[must_use]
+    pub fn internal<E: ToString>(error: E) -> Self {
+        Self {
+            code: error_codes::INTERNAL_ERROR,
+            message: error.to_string(),
+            data: None,
+        }
+    }
+
+    /// Attempt to reparent an entity to itself.
+    #[must_use]
+    pub fn self_reparent(entity: Entity) -> Self {
+        Self {
+            code: error_codes::SELF_REPARENT,
+            message: format!("Cannot reparent Entity {entity} to itself"),
+            data: None,
+        }
+    }
+}
+
+/// Error codes used by BRP.
+pub mod error_codes {
+    /// Invalid JSON.
+    pub const PARSE_ERROR: i16 = -32700;
+
+    /// JSON sent is not a valid request object.
+    pub const INVALID_REQUEST: i16 = -32600;
+
+    /// The method does not exist / is not available.
+    pub const METHOD_NOT_FOUND: i16 = -32601;
+
+    /// Invalid method parameter(s).
+    pub const INVALID_PARAMS: i16 = -32602;
+
+    /// Internal error.
+    pub const INTERNAL_ERROR: i16 = -32603;
+
+    // Bevy errors
+
+    /// Entity not found.
+    pub const ENTITY_NOT_FOUND: i16 = -23401;
+
+    /// Could not reflect or find component.
+    pub const COMPONENT_ERROR: i16 = -23402;
+
+    /// Could not find component in entity.
+    pub const COMPONENT_NOT_PRESENT: i16 = -23403;
+
+    /// Cannot reparent an entity to itself.
+    pub const SELF_REPARENT: i16 = -23404;
+}
+
+/// The result of a request.
+pub type BrpResult = Result<Value, BrpError>;
+
+/// The requests may occur on their own or in batches.
+/// Actual parsing is deferred for the sake of proper
+/// error reporting.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum BrpBatch {
+    /// Multiple requests with deferred parsing.
+    Batch(Vec<Value>),
+    /// A single request with deferred parsing.
+    Single(Value),
+}
+
+/// A message from the Bevy Remote Protocol server thread to the main world.
+///
+/// This is placed in the [`BrpMailbox`].
+#[derive(Debug, Clone)]
+pub struct BrpMessage {
+    /// The request method.
+    pub method: String,
+
+    /// The request params.
+    pub params: Option<Value>,
+
+    /// The channel on which the response is to be sent.
+    ///
+    /// The value sent here is serialized and sent back to the client.
+    pub sender: Sender<BrpResult>,
+}
+
+/// A resource that receives messages sent by Bevy Remote Protocol clients.
+///
+/// Every frame, the `process_remote_requests` system drains this mailbox and
+/// processes the messages within.
+#[derive(Debug, Resource, Deref, DerefMut)]
+pub struct BrpMailbox(Receiver<BrpMessage>);
+
+impl RemotePlugin {
```

Comment:
Nit: Why are some impls next to their types and others away from them? The macro crate for `bevy_reflect` does a similar thing (which I think was my fault haha) and it can be slightly annoying jumping around between definition and implementation.

Reply:
(Because this was basically co-written by three people :P ) That also bothers me, so I'll try to change it.